src/main/java/com/getindata/connectors/http/sink/httpclient/JavaNetSinkHttpClient.java

```diff
@@ -0,0 +1,111 @@
+ */
+public class JavaNetSinkHttpClient implements SinkHttpClient {
+  private final HttpClient httpClient;
+
+  public JavaNetSinkHttpClient() {
+    this.httpClient = HttpClient.newBuilder().followRedirects(HttpClient.Redirect.NORMAL).build();
+  }
+
+  /**
+   * A wrapper structure around an arbitrary element, keeping a reference to a particular
+   * {@link HttpSinkRequestEntry}. Used internally by the {@code HttpSinkWriter} to pass
+   * {@code HttpSinkRequestEntry} along some other element that is logically connected with it
+   * (e.g., full HTTP request built from the {@code HttpSinkRequestEntry}).
+   *
+   * @param <T>
+   */
+  @RequiredArgsConstructor
+  @EqualsAndHashCode
+  private static class HttpSinkRequestEntryWrapper<T> {
+    /**
+     * An element logically connected with the {@link HttpSinkRequestEntry}.
+     */
+    public final T element;
+
+    /**
+     * A representation of a single {@link HttpSink} request.
+     */
+    public final HttpSinkRequestEntry sinkRequestEntry;
+  }
+
+  @Override
+  public CompletableFuture<SinkHttpClientResponse> putRequests(
+      List<HttpSinkRequestEntry> requestEntries, String endpointUrl
+  ) {
+    var endpointUri = URI.create(endpointUrl);
+
+    List<HttpSinkRequestEntryWrapper<HttpRequest>> requests = requestEntries
+        .stream()
+        .map(requestEntry -> new HttpSinkRequestEntryWrapper<>(
+            buildHttpRequest(requestEntry, endpointUri),
+            requestEntry
+        ))
+        .collect(Collectors.toList());
+
+    return getCompletedFutures(requests).thenApply(this::prepareHttpClientResponse);
+  }
+
+  private HttpRequest buildHttpRequest(HttpSinkRequestEntry requestEntry, URI endpointUri) {
+    return HttpRequest
+        .newBuilder()
+        .uri(endpointUri)
+        .version(java.net.http.HttpClient.Version.HTTP_1_1)
+        .header("Content-Type", requestEntry.contentType)
+        .method(requestEntry.method, HttpRequest.BodyPublishers.ofByteArray(requestEntry.element))
+        .build();
+  }
+
+  private CompletableFuture<List<HttpSinkRequestEntryWrapper<Optional<HttpResponse<String>>>>> getCompletedFutures(
+      List<HttpSinkRequestEntryWrapper<HttpRequest>> requests
+  ) {
+    var futures = requests
+        .stream()
+        .map(req -> httpClient
+            .sendAsync(req.element, HttpResponse.BodyHandlers.ofString())
+            .exceptionally(ex -> null)
+            .thenApply(res -> new HttpSinkRequestEntryWrapper<>(Optional.ofNullable(res), req.sinkRequestEntry))
+        )
+        .collect(Collectors.toList());
+
+    var allFutures = CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
+    return allFutures.thenApply(_void -> futures.stream().map(CompletableFuture::join).collect(Collectors.toList()));
+  }
+
+  private SinkHttpClientResponse prepareHttpClientResponse(
+      List<HttpSinkRequestEntryWrapper<Optional<HttpResponse<String>>>> responses
```

Comment:
This is a really long signature, with a long list of generics. maybe we could make it shorter?

1. You are using HttpSinkRequestEntryWrapper for a response, even though the name suggest it is for request.
What is that you really need here?
2. You are passing an Optional as an argument, which forces you to use Optional in Signature for generic, this can be fixed by https://github.com/getindata/flink-http-connector/pull/4/files#r908567620  and https://github.com/getindata/flink-http-connector/pull/4/files#r908560065